### ISSUE TITLE
perf(metalgemm): reuse Qwen Q4_K gate/up output slab across prefill

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#9102]
## Parent context

#8972 and #8697.

## Current state

`internal/metalgemm.GEMMGroupWithEvents` allocates `ycat := make([]float32, off)` for every grouped Q4_K call. Qwen resident-Q4_K prefill invokes it for every compute layer's gate/up pair.

The exact checkpoint declares 65 blocks but one is a NextN/MTP draft block; the loader correctly exposes 64 compute layers. At the closed #9083 ceiling P=512 and exact I=17,408, each gate/up pair allocates 71,303,168 bytes (68 MiB), or 4.25 GiB over 64 layers per full chunk. Across the frozen P=32,800 suffix (64 full chunks plus one 32-token tail), this is 292,342,988,800 bytes (272.266 GiB) of cumulative allocation traffic although only one layer consumes the backing at a time. Closed #9094 observed 12,671.94 MiB swap growth before a response.

## Why this is next

Width-only tuning is exhausted: P=512 survives but misses the endpoint deadline, while P=1024 is unsafe. Reusing the already-serial gate/up readback backing is the smallest disjoint allocation-ownership lever before another width change or broad resident graph.

## Priority context

Working path: bounded P=512 chunks -> one session-local grouped-output backing -> deterministic parity/lifecycle -> exact Mac A/B -> resume #8972.

Current blocker: repeated 68 MiB call-owned gate/up allocations amplify Go heap and GC pressure on every layer and chunk.

Unblocks: a lower-pressure prefill control for #8972 and evidence for #8697.

Not polish: the exact long request cannot currently produce a response within its safety/time envelope.

## Working spine

Add a caller-buffer grouped Q4_K GEMM form, retain the allocating API as fallback, let each `Session` lazily own one P<=512 output slab, reuse it across sequential groups and chunks, and clear it on `Session.Close`.

## Value frame

For: M3 Pro operators serving exact Qwen3.8-27B Q4_K_M through fak-native Metal.

Problem: each layer and chunk allocates and garbage-collects another 68 MiB gate/up output.

Today: keep the safe 512-token ceiling and pay 272.266 GiB cumulative allocation on the frozen suffix.

Better because: one bounded per-session backing removes repeated heap growth without changing Metal math, readback, chunk width, or execution ownership.

Witness: one readback test binds backing reuse, lifecycle, parity, native identity, and the Mac verdict.

## Problem frame

Centrality: Core - changes request-time pressure on the native path blocking #8972.

P1 Context: preserved - prompt, KV/recurrent state, order, logits, and generated tokens do not change.

P2 Net value: advanced - allocation reuse, GC, footprint/swap, and TTFT are gated together.

P3 Adaptation: advanced - reuse is lazy, session-local, capped to P<=512, and larger or unsupported panels retain the allocating fallback.

P4 Operations: advanced - identity, high-water, teardown, and KEEP/REJECT are captured as immutable evidence.

## In scope

Add one grouped-Q4_K into-caller-buffer API; one lazy per-session host-f32 slab capped at P<=512; route existing grouped Metal prefill through it; clear it on close; and add shape, aliasing, isolation, lifecycle, parity, allocation, and exact-Mac A/B evidence.

## Out of scope

No shader math, command-buffer-count change, Q8/Q6_K, GDN ownership (#9097), KV reservation (#9094), chunk policy (#9089), mixed full-attention QKV (#8833), decode, general/global allocator, llama.cpp execution, or slab retention above P=512. Do not claim long-request completion or #8697 parity.

Priority, shared-trunk coordination, and runtime feature gating remain orthogonal.

## Done condition

One resolving commit makes supported P<=512 fak-native Qwen grouped-Q4_K prefill reuse one session-owned backing across all sequential gate/up layer calls and later chunks, preserves outputs and native identity, isolates sessions, releases backing on close, keeps the safe fallback outside the envelope, and records KEEP only if the Mac gate passes or REJECT with no default change otherwise.

## Witness

`go test ./docs/_witnesses/issue-9102-qwen38-gateup-slab -run '^TestGateUpSlabWitness$' -count=1 -v`

## Acceptance gate

Focused portable and Darwin tests prove caller-buffer bounds, supplied-slab aliasing, one stable backing across 64 gate/up calls and a second chunk, session isolation, close teardown, untouched fallback, and parity with allocating and CPU controls.

The hardware arm uses control and candidate builds on the same 36 GiB, 18-GPU-core M3 Pro; artifact SHA-256 `7e78da5d7e3ae28d178121f58646953305f3e5bd3cb46f4a75584e8b6c6fe169`; fak-native Metal/Q4_K identity; identical temperature-zero P=512/T=8 input; fresh sessions; three repetitions per arm; and exclusive machine ownership.

KEEP requires identical accepted tokens, deterministic proof that one backing retires 63 allocations per full-layer traversal, no increase in GC cycles, peak footprint, or swap, median TTFT no worse by more than 2%, and at least one of GC cycles, peak footprint, or median TTFT strictly better. Otherwise record REJECT and leave reuse disabled.

## Dedupe

All-state searches on 2026-08-26 found no issue owning this exact session-local grouped-output allocation.

- Closed #1113 is the broad prefill-roundtrip predecessor, but its bulk-upload and grouped-GEMM work left literal `make(ycat)` present.
- Open #8833 explicitly excludes MLP.
- Closed #9083 excludes Metal kernels.
- Closed #9094 owns KV reservation.
- Open #9097 owns resident GDN sequence integration.
- Open #8101 owns load/startup materialization.

## Dependencies

- Builds on: #9083.
- Related: #8972, #8697, #1113, #9097.

## Risks

A retained 68 MiB slab per live session can amplify concurrency, so it must be lazy, capped, session-local, and cleared on close.

Returned G/U slices alias the slab and must be fully consumed before the next grouped call.

P>512 must retain the existing call-owned allocation rather than retaining a larger slab.

## Closure binding

Close only from one signed commit carrying `Fixes #9102` and `(fak metalgemm)`, implementation, focused tests, and the immutable Mac packet; passing exact-path `fak validate --mine` and `dos commit-audit`; and recording the declared KEEP with reuse enabled or durable REJECT with the default unchanged.

## Lane

qwen38-gemm-slab

## Likely files

- `internal/metalgemm/q4k.go`
- `internal/metalgemm/q4k_test.go`
- `internal/model/kv.go`
- `internal/model/hal.go`
- `internal/model/metal_q4k_on.go`
- `internal/model/metal_q4k_test.go`
- `docs/_witnesses/issue-9102-qwen38-gateup-slab/`

## Work unit

leaf

## Expected steps

5

## Work scale

S1

## Required model tier

tier/T2-required

## Optimal model tier

tier/T1-optimal

## Assumptions

- The exact artifact retains 64 compute layers after excluding its single declared NextN/MTP block.
- One Session executes prefill groups serially.
- The Mac packet can isolate request-time GC and process footprint from model load.
- Invalidating assumption: downstream work may retain grouped output across the next grouped dispatch.

## Confusion risks

- This is a host Go f32 readback slab, not device-resident activation fusion.
- Lower cumulative allocation alone does not prove lower peak footprint or TTFT.
- This does not replace #9097 or complete P=32,800.

## Coordination

- Acquire the exact keyword lease before edits and the machine-wide GPU lease only for the Mac arm.
- The proposed paths were admitted by `dos arbitrate` against live #9097/#8385 leases; do not add `qwen35_prefill_q4k.go` or another leased path without re-arbitration.
- Serialize Darwin tests and the live A/B with other Metal workers, then restore the comparator by exact command identity.

## Trigger

The #8972 P=1024 verdict selected allocation ownership and resident amplification after width tuning became unsafe; source inspection localizes the largest repeated allocation to gate/up `ycat`.

## Batch policy

One issue per marker `qwen38-q4k-gateup-slab-2026-08-26`; reruns update it. Do not file per-layer, per-chunk, allocator, or hardware-result duplicates.

## Generation

- Stream: gen/now.
- Promotion evidence: #8972 selected allocation ownership after P=1024 became unsafe, and current source proves a 68 MiB per-layer allocation.
- Demotion evidence: retire with REJECT if reuse cannot improve GC, footprint, or TTFT without regression.
- Invalidating assumption: downstream work may retain grouped output across the next grouped dispatch.
- Orthogonality: generation does not change P1 priority, default enablement, lease rules, or parent parity criteria.

## Target operating envelope

- unified memory: = 36 gib
- gpu cores: = 18 cores
- compute layers: = 64
- prefill panel: = 512 tokens
- intermediate width: = 17408 elements
- grouped output high water: <= 71303168 bytes

## Witnessed operating envelope

- unified memory: 36 gib
- gpu cores: 18 cores
- compute layers: 64
- prefill panel: 512 tokens
- intermediate width: 17408 elements
- grouped output high water: 71303168 bytes

## Completion standard

development

## Work estimate

Estimate: 3 points. Uncertainty: request-scoped GC capture may need a narrow existing profiler hook, not a general telemetry subsystem.

## Overall completion contribution

Contribution: 3/3 points for this bounded leaf; long-request completion and comparator parity remain in #8972/#8697.

**Comment by anthony-chaudhary** [comment]
Dispatch claim (2026-08-26 PT): all-state dedupe found no owner for this exact session-local grouped-output allocation; #1113 left the literal ycat allocation, while #8833/#9097/#9094 own QKV/GDN/KV work. The corrected exact geometry is 64 compute layers after the GGUF single NextN block is excluded: 68 MiB per gate/up pair, 4.25 GiB per full P=512 chunk, and 272.266 GiB cumulative traffic over P=32,800. The one-deliverable/one-witness lint is PASS, and DOS admits the exact qwen38-gemm-slab tree alongside active #9097/#8385 leases. I am claiming only deterministic implementation now; the live Mac A/B remains serialized under the machine GPU lease and KEEP/REJECT defaults remain evidence-bound.

**Comment by anthony-chaudhary** [comment]
Closed on pushed `main` with the conservative Mac verdict **REJECT**; reuse remains default-off.

- `89b5566caf44` adds the caller-buffer grouped Q4_K path, one bounded session-local MLP-output backing, close-time allocation/reuse/high-water receipt, and focused parity/lifecycle/fallback tests. `FAK_Q4K_GATEUP_SLAB=1` is still required.
- `3e5891679d4b` binds the immutable control response, all 132 raw memory samples, exact native identity, zero candidate execution, watcher/teardown, and 90-second comparator restoration.
- The pinned default-off control completed fak-native Metal/Q4_K P=512/T=8, but samples 129–131 crossed the predeclared first-sample swap +12 GiB boundary before response completion. The candidate was therefore never admitted; there is no speed, GC, or safety improvement claim.
- Exact committed-tip Metal/model/fallback/artifact tests pass, and `dos commit-audit 89b5566c^..3e589167` reports both commits `OK` / `diff-witnessed`.

The closure binding expected code and packet in one commit. The sanctioned isolated lander omitted the initially untracked directory and reported a post-CAS pathspec sync failure, so the proof was recovered into a fresh sanctioned worktree and pushed as the immediately adjacent commit in the same push. That tooling defect is now durably tracked as #9129; no remote state ever contained the implementation without its packet.
